### ISSUE TITLE
Very slow on CPU

### ISSUE THREAD

**Issue by tzuf** [pytorch/pytorch#4703]
my model was really slow with pytorch so I tried one of your examples and saw it is slow as well ([seq2seq](http://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html)).

It takes a long time...:
27m 55s (- 390m 53s) (5000 6%) 2.8939
56m 31s (- 367m 27s) (10000 13%) 2.3238
85m 20s (- 341m 23s) (15000 20%) 2.0054
117m 23s (- 322m 50s) (20000 26%) 1.7787
146m 48s (- 293m 36s) (25000 33%) 1.5661

I’m running only on CPU
I’m working on Ubuntu 16.04.3.
64 cores.
Python 3.6.3 :: Anaconda custom (64-bit).

tried using MKL (installed via conda)
didn't help.

Any ideas?

**Comment by albanD** [comment]
Hi,

Could you try limiting the number of available cores to pytorch?
You can do so by calling [`torch.set_num_threads()`](http://pytorch.org/docs/0.3.0/torch.html#torch.get_num_threads) and/or using the env variable `OMP_NUM_THREADS`.

**Comment by tzuf** [comment]
Hi,
I tried setting OMP_NUM_THREADS to 1, 4 and 8.
It helped a lot but it was still very slow

**Comment by albanD** [comment]
CPU computation is expected to be slow so it is most certainly normal.
Could you quantify "very slow" and why do you think it is abnormally slow?

**Comment by tzuf** [comment]
After changing OMP_NUM_THREADS  it still takes the seq2seq tutorial more than 2 hours. It's not a lot of data and the model  so it seems to me to be too slow. Also when I ran my model and data with theano it was much faster although I ran it on CPU as well

**Comment by fmassa** [comment]
Could you print the version of pytorch that you are using?
```python
print(torch.__version__)
```
Also, there is a PR with some speed improvements for the CPU will be merged soon https://github.com/pytorch/pytorch/pull/2764

**Comment by tzuf** [comment]
the version is: 0.3.0

**Comment by MlWoo** [comment]
@tzuf Could you provide the CPU type, you benchmark and the commit id of PyTorch?

**Comment by tzuf** [comment]
the cpu: Intel(R) Xeon(R) CPU           X7560  @ 2.27GHz
I'm working on the data set from pytorch tutorial [(seq2seq)](http://pytorch.org/tutorials/intermediate/seq2seq_translation_tutorial.html).
The version is 0.3.0. don't know commit id

**Comment by MlWoo** [comment]
@tzuf My colleague help me try to reproduce the problems on Intel(R) Core(TM) i7-6700 CPU @ 3.40GHz whose performance is even worse than that on your machine. Unfortunately, it only costs 3min (not nearly 27 min) to print one log. Could you tell me which demo dataset and scripts that you use?
I think it maybe caused by your environments. You can use command `top` to check the utilization of CPU. 

Btw, there is 8 cores(not 64 cores that you mentioned) on Intel(R) Xeon(R) CPU X7560.   `OMP_NUM_THREADS=8 KMP_AFFINITY=scatter` is a good chioce.

**Comment by juliusshufan** [comment]
@tzuf 
Hi, 

I run the "seq2seq_translation_tutorial.py" with dataset downloading from http://www.manythings.org/anki/fra-eng.zip
My CPU model is i7-6700 CPU @ 3.40GHz, and which contains 4 physical cores.

I installed official PyTorch v0.3 via conda, by command "conda install pytorch-cpu torchvision -c pytorch", and setting OMP_NUM_THREADS=4 before running, looks like it faster than yours, see logs below:
Reading lines...
Read **145437** sentence pairs
Trimmed to 11318 sentence pairs
Counting words...
Counted words:
fra **4589**
eng **2988**
['je suis la plus jeune enfant .', 'i m the youngest child .']
3m 19s (- 46m 27s) (5000 6%) 2.9148
6m 37s (- 43m 1s) (10000 13%) 2.3547
9m 59s (- 39m 57s) (15000 20%) 2.0318
13m 18s (- 36m 36s) (20000 26%) 1.7777
16m 38s (- 33m 17s) (25000 33%) 1.5810
19m 58s (- 29m 57s) (30000 40%) 1.4056
23m 18s (- 26m 37s) (35000 46%) 1.2725
26m 38s (- 23m 18s) (40000 53%) 1.1402
29m 56s (- 19m 57s) (45000 60%) 1.0574

Can you please clarifiy the dataset you using, at the size of the dataset will also compact the time cost.
Thanks.

**Comment by tzuf** [comment]
Hi I did as you did:
1. downloaded the same data-set and
2. I installed official PyTorch v0.3 via conda, by command "conda install pytorch-cpu torchvision -c pytorch", 
3. set OMP_NUM_THREADS=4

It takes me double the time:

Reading lines...
Read 145437 sentence pairs
Trimmed to 11318 sentence pairs
Counting words...
Counted words:
fra 4589
eng 2988
['tu es fort effronte .', 'you re very forward .']
6m 28s (- 90m 42s) (5000 6%) 2.9297
13m 6s (- 85m 9s) (10000 13%) 2.3578
19m 43s (- 78m 54s) (15000 20%) 2.0613
26m 17s (- 72m 18s) (20000 26%) 1.8068
32m 48s (- 65m 37s) (25000 33%) 1.6223
38m 58s (- 58m 27s) (30000 40%) 1.4049
44m 22s (- 50m 42s) (35000 46%) 1.3244
49m 46s (- 43m 33s) (40000 53%) 1.1715
55m 15s (- 36m 50s) (45000 60%) 1.0675
60m 35s (- 30m 17s) (50000 66%) 0.9961
65m 57s (- 23m 58s) (55000 73%) 0.9050
71m 23s (- 17m 50s) (60000 80%) 0.8221
76m 55s (- 11m 50s) (65000 86%) 0.7823
82m 24s (- 5m 53s) (70000 93%) 0.7406
87m 42s (- 0m 0s) (75000 100%) 0.6467

**Comment by tzuf** [comment]
@MlWoo there are 64 cores (I see them in htop and also the administrator told me).
Which information do you need? 
After limiting OMP_NUM_THREADS=4 the CPU usage is at around 400%

**Comment by MlWoo** [comment]
@tzuf The [official site] (https://ark.intel.com/products/46499/Intel-Xeon-Processor-X7560-24M-Cache-2_26-GHz-6_40-GTs-Intel-QPI) shows the info. Is your machine a cluster of CPU?

**Comment by tzuf** [comment]
@MlWoo 
Tried OMP_NUM_THREADS=8 got the same results as 4 . I'm checking the seq2seq tutorial. According to the administrator, it is not a cluster.

Reading lines...
Read 145437 sentence pairs
Trimmed to 11318 sentence pairs
Counting words...
Counted words:
fra 4589
eng 2988
['vous etes pleins de ressources .', 'you re very resourceful .']
6m 41s (- 93m 41s) (5000 6%) 2.9300
13m 18s (- 86m 29s) (10000 13%) 2.3453
19m 58s (- 79m 55s) (15000 20%) 2.0239
26m 33s (- 73m 2s) (20000 26%) 1.8138
33m 9s (- 66m 18s) (25000 33%) 1.5948
39m 8s (- 58m 43s) (30000 40%) 1.4337
45m 53s (- 52m 26s) (35000 46%) 1.3120
52m 28s (- 45m 55s) (40000 53%) 1.1779
59m 11s (- 39m 27s) (45000 60%) 1.0523
65m 49s (- 32m 54s) (50000 66%) 0.9511
72m 21s (- 26m 18s) (55000 73%) 0.8931
79m 5s (- 19m 46s) (60000 80%) 0.8126
85m 50s (- 13m 12s) (65000 86%) 0.7484
92m 23s (- 6m 35s) (70000 93%) 0.6700
99m 4s (- 0m 0s) (75000 100%) 0.6148

**Comment by nyxjemk** [comment]
Any news on this?
I'm having the same problem. PyTorch is very slow on CPU, it's also only using 2-3 cores with 16 cores available. With TF it is no problem to use ~8-10 cores. Which also results in much better performance.

Setting any value for OMP_NUM_THREADS makes it only worse.

**Comment by apaszke** [comment]
@nyxjemk are you also having trouble with the same tutorial?

**Comment by nyxjemk** [comment]
I was trying this LSTM example (and some others from the repository):
https://github.com/yunjey/pytorch-tutorial/blob/master/tutorials/02-intermediate/recurrent_neural_network/main.py



I tried building pytorch from source. Somehow it was extremely slow after this (I'm using an virtual server with vCores, maybe the script got confused by virtual hardware?) so I went back to using the pip installation (0.3.1). 
When using two threads, performance seems to be slightly better than default.

**Comment by MlWoo** [comment]
@nyxjemk I am sorry that it is really hard to identify the root because we do not have the virtual env like yours.  LSTM mainly involves with BLAS which is usually accelerated by MKL or openBLAS. I just identify a severe bug #5047 of BLAS usage in the recent version. Could you provide the commit id of the source?  And I will try to reproduce your problem and identify the root cause.

**Comment by CircleXing001** [comment]
emmm.... I am sorry that i ran the same data and model(lstm) on the same cpu with pytorch and tf but pytorch is much slower than tf.I don't know why it does  this .Can you help me?THX

**Comment by CircleXing001** [comment]
@MlWoo I installed pytorch from binaries.I tried a simple LSTM but it was 10x slower than tf.Thanks for your reply.

**Comment by MlWoo** [comment]
You can refer to [the setting](https://github.com/intel/pytorch#bkm-on-xeon). If you are not sure the number of threads, just set the env variables `export KMP_AFFINITY=granularity=fine,compact,1,0`.

**Comment by CircleXing001** [comment]
@MlWoo I tried，but it didn't dobetter.Here is my code .
hidden_num=50
layer_num=1
time_step=1
input_num=8
output_num=1
epoch_num=50
batch_size=72
learning_rate=0.001
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.lstm1=nn.LSTM(input_size=input_num,hidden_size=hidden_num,num_layers=layer_num)
        self.fc1=nn.Linear(hidden_num,output_num)
    def forward(self,x):
        out,_=self.lstm1(x)
        x=self.fc1(out[:,-1,:])
        x=x.view(-1,output_num)
        return x
model=Net()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
loss_epoch=[0]*epoch_num
def train():
    model.train()
    xx=0
    n=0
    while train_XX._epochs_completed<epoch_num:
        batch_X=train_XX.next_batch(batch_size)
        batch_y=train_yy.next_batch(batch_size)
        batch_X=torch.from_numpy(batch_X)
        batch_y=torch.from_numpy(batch_y)
        optimizer.zero_grad()
        output=model(batch_X)
        loss=F.l1_loss(output,batch_y)
        loss.backward()
        optimizer.step()
        if train_XX._epochs_completed==xx:
            n=n+1
            loss_epoch[xx]+=float(loss)
        else:
            loss_epoch[xx]=loss_epoch[xx]/n
            xx=train_XX._epochs_completed
            n=1
            print(loss_epoch[xx-1])
            if xx!=epoch_num:
                loss_epoch[xx]+=float(loss)
train()

**Comment by CircleXing001** [comment]
@MlWoo I tried,but it didn't do better .Here is my code.
`hidden_num=50
layer_num=1
time_step=1
input_num=8
output_num=1
epoch_num=50
batch_size=72
learning_rate=0.001
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.lstm1=nn.LSTM(input_size=input_num,hidden_size=hidden_num,num_layers=layer_num)
        self.fc1=nn.Linear(hidden_num,output_num)
    def forward(self,x):
        out,_=self.lstm1(x)
        x=self.fc1(out[:,-1,:])
        x=x.view(-1,output_num)
        return x
model=Net()
optimizer = optim.Adam(model.parameters(), lr=learning_rate)
loss_epoch=[0]*epoch_num
def train():
    model.train()
    xx=0
    n=0
    while train_XX._epochs_completed<epoch_num:
        batch_X=train_XX.next_batch(batch_size)
        batch_y=train_yy.next_batch(batch_size)
        batch_X=torch.from_numpy(batch_X)
        batch_y=torch.from_numpy(batch_y)
        optimizer.zero_grad()
        output=model(batch_X)
        loss=F.l1_loss(output,batch_y)
        loss.backward()
        optimizer.step()
        if train_XX._epochs_completed==xx:
            n=n+1
            loss_epoch[xx]+=float(loss)
        else:
            loss_epoch[xx]=loss_epoch[xx]/n
            xx=train_XX._epochs_completed
            n=1
            print(loss_epoch[xx-1])
            if xx!=epoch_num:
                loss_epoch[xx]+=float(loss)
train()       
`

**Comment by MlWoo** [comment]
@CircleXing001 there is a pen on the top-right of your conversation box. you can use it to modify the content. But now the format of your code is still in chaos.

**Comment by CircleXing001** [comment]
@MlWoo I tried but it didn't do better.Here is my code.yeah, i'm sorry about it.so i upload it,hh
[code.txt](https://github.com/pytorch/pytorch/files/1979003/code.txt)
